src/buffer/out/textBuffer.cpp

```diff
@@ -950,55 +1018,34 @@ void TextBuffer::Reset()
-        til::CoordType TopRow = 0; // new top row of the screen buffer
-        if (newSize.height <= GetCursor().GetPosition().y)
+        TextBuffer newBuffer{ newSize, _currentAttributes, 0, false, _renderer };
+        const auto cursorRow = GetCursor().GetPosition().y;
+        const auto copyableRows = std::min<til::CoordType>(_height, newSize.height);
+        til::CoordType srcRow = 0;
+        til::CoordType dstRow = 0;
+
+        if (cursorRow >= newSize.height)
         {
-            TopRow = GetCursor().GetPosition().y - newSize.height + 1;
+            srcRow = cursorRow - newSize.height + 1;
         }
-        const auto TopRowIndex = gsl::narrow_cast<size_t>(_firstRow + TopRow) % _storage.size();
-
-        std::vector<ROW> newStorage;
-        auto newBuffer = _allocateBuffer(newSize, _currentAttributes, newStorage);
 
-        // This basically imitates a std::rotate_copy(first, mid, last), but uses ROW::CopyRangeFrom() to do the copying.
+        for (; dstRow < copyableRows; ++dstRow, ++srcRow)
         {
-            const auto first = _storage.begin();
-            const auto last = _storage.end();
-            const auto mid = first + TopRowIndex;
-            auto dest = newStorage.begin();
-
-            std::span<ROW> sourceRanges[]{
-                { mid, last },
-                { first, mid },
-            };
-
-            // Ensure we don't copy more from `_storage` than fit into `newStorage`.
-            if (sourceRanges[0].size() > newStorage.size())
-            {
-                sourceRanges[0] = sourceRanges[0].subspan(0, newStorage.size());
-            }
-            if (const auto remaining = newStorage.size() - sourceRanges[0].size(); sourceRanges[1].size() > remaining)
-            {
-                sourceRanges[1] = sourceRanges[1].subspan(0, remaining);
-            }
-
-            for (const auto& sourceRange : sourceRanges)
-            {
-                for (const auto& oldRow : sourceRange)
-                {
-                    til::CoordType begin = 0;
-                    dest->CopyRangeFrom(0, til::CoordTypeMax, oldRow, begin, til::CoordTypeMax);
-                    dest->TransferAttributes(oldRow.Attributes(), newSize.width);
-                    ++dest;
-                }
-            }
+            newBuffer.GetRowByOffset(dstRow).CopyFrom(GetRowByOffset(srcRow));
         }
 
-        _charBuffer = std::move(newBuffer);
-        _storage = std::move(newStorage);
+        // NOTE: Keep this in sync with _reserve().
+        _buffer = std::move(newBuffer._buffer);
+        _bufferEnd = newBuffer._bufferEnd;
+        _commitWatermark = newBuffer._commitWatermark;
+        _initialAttributes = newBuffer._initialAttributes;
+        _bufferRowStride = newBuffer._bufferRowStride;
+        _bufferOffsetChars = newBuffer._bufferOffsetChars;
+        _bufferOffsetCharOffsets = newBuffer._bufferOffsetCharOffsets;
+        _width = newBuffer._width;
+        _height = newBuffer._height;

```

Comment:
curious, why not `std::move` all of these over?

Reply:
Most of them are POD. moving POD doesn't do anything! you move when you want to take full ownership of _an actual resource_ that something else is holding onto. it's great for a vector, for example: it steals the memory allocation out of the old vector!